src/main/scala/org/allenai/pnp/BsoTrainer.scala

```diff
@@ -0,0 +1,217 @@
+package org.allenai.pnp
+
+import scala.collection.mutable.ListBuffer
+
+import com.google.common.base.Preconditions
+import com.jayantkrish.jklol.training.LogFunction
+import com.jayantkrish.jklol.util.KbestQueue
+
+import edu.cmu.dynet._
+import edu.cmu.dynet.dynet_swig._
+
+/**
+ * Beam search trainer implementing a LaSO-like algorithm.
+ * This trainer implements a margin loss between the
+ * lowest-scoring item on the beam and the highest-scoring
+ * correct execution. Thus, a gradient update is performed
+ * every time all correct executions fall off the beam.
+ */
+class BsoTrainer(val epochs: Int, val beamSize: Int, val maxIters: Int,
+    val model: PpModel, val trainer: Trainer, val log: LogFunction) {
+
+  Preconditions.checkArgument(model.locallyNormalized == false)
+
+  import DynetScalaHelpers._
+
+  def train[A](examples: Seq[PpExample[A]]): Unit = {
+    for (i <- 0 until epochs) {
+      var loss = 0.0
+      log.notifyIterationStart(i)
+      for (example <- examples) {
+        loss +=  doExampleUpdate(example)
+      }
+      
+      trainer.update_epoch()
+
+      log.logStatistic(i, "loss", loss)
+      log.notifyIterationEnd(i)
+    }
+  }
+  
+  private def doExampleUpdate[A](example: PpExample[A]): Double = {
+    val cg = new ComputationGraph
+    var loss = 0.0
+
+    val env = example.env
+    val stateCost = if (example.conditionalExecutionScore == null) {
+      ExecutionScore.zero
+    } else {
+      example.conditionalExecutionScore
+    }
+    val graph = model.getComputationGraph(cg)
+
+    val queue = new BsoPpQueue[A](beamSize, stateCost, graph, log)
+    val finished = new BsoPpQueue[A](beamSize, stateCost, graph, log)
+
+    val startEnv = env.setLog(log)
+    queue.offer(example.unconditional, env, 0.0, null, null, env)
+
+    log.startTimer("bso/beam_search")
+    val beam = new Array[SearchState[A]](beamSize)
+    var numIters = 0
+    val losses = ListBuffer[Expression]()
+    while (queue.queue.size > 0 &&
```

Comment:
The queue has several different queues in it, each of which has its own size, so I think this way makes sense. We're never going to use the class outside this file.